zebra-state/src/service/write.rs

```diff
@@ -108,31 +176,105 @@ pub fn write_blocks_from_channels(
 
-        // TODO:
-        // - read from the channel
-        // - commit blocks to the non-finalized state
-        // - if there are any ready, commit blocks to the finalized state
-        // - handle errors by sending a reset with all the block hashes in the non-finalized state, and the finalized tip
-        // - update the chain tip sender and cached non-finalized state
-        error!("handle non-finalized block writes here");
-    }
+    while let Some(non_finalized_write_cmd) = non_finalized_block_write_receiver.blocking_recv() {
+        match non_finalized_write_cmd {
+            NonFinalizedWriteCmd::ProcessQueued {
+                parent_hash,
+                queued_child: (child, rsp_tx),
+            } => {
+                let parent_error = parent_error_map.get(&parent_hash);
+
+                let child_hash = child.hash;
+                let result;
+
+                // If the block is invalid, reject any descendant blocks.
+                //
+                // At this point, we know that the block and all its descendants
+                // are invalid, because we checked all the consensus rules before
+                // committing the block to the non-finalized state.
+                // (These checks also bind the transaction data to the block
+                // header, using the transaction merkle tree and authorizing data
+                // commitment.)
+                if let Some(parent_error) = parent_error {
+                    tracing::trace!(
+                        ?child_hash,
+                        ?parent_error,
+                        "rejecting queued child due to parent error"
+                    );
+                    result = Err(parent_error.clone());
+                } else {
+                    tracing::trace!(?child_hash, "validating queued child");
+                    result = validate_and_commit_non_finalized(
+                        &finalized_state,
+                        &mut non_finalized_state,
+                        network,
+                        child,
+                    )
+                    .map_err(CloneError::from);
+
+                    if result.is_ok() {
+                        // Update the metrics if semantic and contextual validation passes
+                        metrics::counter!("state.full_verifier.committed.block.count", 1);
+                        metrics::counter!("zcash.chain.verified.block.total", 1);
+                    }
+                }
+
+                let _ = rsp_tx.send(result.clone().map(|()| child_hash).map_err(BoxError::from));
+
+                if let Err(ref error) = result {
+                    parent_error_map.insert(parent_hash, error.clone());
+                }
+            }
+
+            NonFinalizedWriteCmd::FinishProcessingQueued => {
+                // Clear out any errors from processed queued blocks without children
+                parent_error_map.clear();
```

Comment:
That does feel like a wasted opportunity to avoid `validate_and_commit` calls when children that need to be failed aren't in the queue yet. 

I'll update it once the tests are passing since this is closer to the pre-existing logic (it's currently sending the FinishProcessingQueued message where it was previously dropping the errors of any parent blocks without queued children.)

Reply:
Updated in [`a9d4e05`](https://github.com/ZcashFoundation/zebra/pull/5257/commits/a9d4e0571959aa8aab067148840578445aff1147), let me know if that's right.

Reply:
Looks good, I simplified the truncation, changed the max size, and added a constant in 0e37923 in PR #5299.